references/embedding/train.py

```diff
@@ -0,0 +1,149 @@
+
+        embeddings = model(samples)
+
+        loss, frac_pos_triplets = criterion(embeddings, targets)
+        loss.backward()
+        optimizer.step()
+
+        running_loss += loss.item()
+        running_frac_pos_triplets += float(frac_pos_triplets)
+
+        if i % print_freq == print_freq - 1:
+            i += 1
+            avg_loss = running_loss / print_freq
+            avg_trip = 100.0 * running_frac_pos_triplets / print_freq
+            print('[{:d}, {:d}] | loss: {:.4f} | % avg hard triplets: {:.2f}%'.format(epoch, i, avg_loss, avg_trip))
+            running_loss = 0
+            running_frac_pos_triplets = 0
+
+
+@torch.no_grad()
+def evaluate(model, dataset, device):
+    model.eval()
+    embeds, labels = None, None
+    dists, targets = None, None
+
+    loader = DataLoader(dataset, batch_size=512, shuffle=False, num_workers=4)
+    for data in loader:
+        samples, _labels = data[0].to(device), data[1]
+        out = model(samples)
+        embeds = torch.cat((embeds, out), dim=0) if embeds is not None else out
+        labels = torch.cat((labels, _labels), dim=0) if labels is not None else _labels
+
+    dists = torch.cdist(embeds, embeds)
+
+    labels = labels.unsqueeze(0)
+    targets = labels == labels.t()
+
+    mask = torch.ones(dists.size()).triu() - torch.eye(dists.size(0))
+    dists = dists[mask == 1]
+    targets = targets[mask == 1]
+
+    best_thresh = 0.01
+    best_correct = 0
+    for thresh in torch.arange(0.0, 1.51, 0.01):
+        predictions = dists <= thresh.to(device)
+        correct = torch.sum(predictions == targets.to(device)).item()
+        if correct > best_correct:
+            best_thresh = thresh
+            best_correct = correct
+
+    accuracy = 100.0 * best_correct / dists.size(0)
+    print('accuracy: {:.3f}%, threshold: {:.2f}'.format(accuracy, best_thresh))
+
+
+def save(model, epoch, save_dir, file_name):
+    file_name = 'epoch_' + str(epoch) + '__' + file_name
+    save_path = os.path.join(save_dir, file_name)
+    torch.save(model.state_dict(), save_path)
+
+
+def main(args):
+    device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
+    p = args.labels_per_batch
+    k = args.samples_per_label
+    batch_size = p * k
+
+    model = EmbeddingNet()
+    if args.resume:
+        model.load_state_dict(torch.load(args.resume))
+
+    model.to(device)
+
+    criterion = TripletMarginLoss(margin=args.margin)
+    optimizer = Adam(model.parameters(), lr=args.lr)
+
+    transform = transforms.Compose([transforms.Lambda(lambda image: image.convert('RGB')),
+                                    transforms.Resize((224, 224)),
+                                    transforms.ToTensor()])
+
+    train_dataset = FashionMNIST(args.train_data, train=True, transform=transform, download=True)
+    test_dataset = FashionMNIST(args.test_data, train=False, transform=transform, download=True)
+
+    targets = train_dataset.targets.tolist()
```

Comment:
A comment here would be helpful, as this is generally something that the user will need to change if they change the dataset

Reply:
I added comments mentioning that any classification dataset should be fine here as long as `targets` is constructed as described. Is that sufficient?

Reply:
The comment saying that it should be any classification dataset is misleading, because not all datasets have the `.targets` attribute, even if they are classification datasets. Maybe just check that the dataset has a `targets` attribute, and raise a nice error message if not?